golem-component-service-base/src/repo/component.rs

```diff
@@ -0,0 +1,510 @@
+
+    async fn get_namespaces(&self, component_id: &Uuid) -> Result<Vec<String>, RepoError>;
+
+    async fn delete(&self, namespace: &str, component_id: &Uuid) -> Result<(), RepoError>;
+}
+
+pub struct DbComponentRepo<DB: Database> {
+    db_pool: Arc<Pool<DB>>,
+}
+
+impl<DB: Database> DbComponentRepo<DB> {
+    pub fn new(db_pool: Arc<Pool<DB>>) -> Self {
+        Self { db_pool }
+    }
+}
+
+#[async_trait]
+impl ComponentRepo for DbComponentRepo<sqlx::Sqlite> {
+    async fn upsert(&self, component: &ComponentRecord) -> Result<(), RepoError> {
+        let mut transaction = self.db_pool.begin().await?;
+
+        let result = sqlx::query(
+            "SELECT count(component_id) as count FROM components WHERE namespace != $1 AND component_id = $2",
+        )
+        .bind(component.namespace.clone())
+        .bind(component.component_id)
+        .fetch_one(&mut *transaction)
+        .await?;
+
+        let count: i64 = result.get("count");
+
+        if count > 0 {
+            return Err(RepoError::Internal("Component id not unique".to_string()));
+        }
+
+        sqlx::query(
+            r#"
+              INSERT INTO components
+                (namespace, component_id, name)
+              VALUES
+                ($1, $2, $3)
+              ON CONFLICT (namespace, name) DO NOTHING
+               "#,
+        )
+        .bind(component.namespace.clone())
+        .bind(component.component_id)
+        .bind(component.name.clone())
+        .execute(&mut *transaction)
+        .await?;
+
+        sqlx::query(
+            r#"
+              INSERT INTO component_versions
+                (component_id, version, size, user_component, protected_component, protector_version, metadata)
+              VALUES
+                ($1, $2, $3, $4, $5, $6, $7)
+              ON CONFLICT (component_id, version) DO UPDATE
+              SET size = $3,
+                  user_component = $4,
+                  protected_component = $5,
+                  protector_version = $6,
+                  metadata = $7
+               "#,
+        )
+            .bind(component.component_id)
+            .bind(component.version)
+            .bind(component.size)
+            .bind(component.user_component.clone())
+            .bind(component.protected_component.clone())
+            .bind(component.protector_version)
+            .bind(component.metadata.clone())
+            .execute(&mut *transaction)
+            .await?;
+
+        transaction.commit().await?;
+
+        Ok(())
+    }
+
+    async fn get(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Vec<ComponentRecord>, RepoError> {
+        sqlx::query_as::<_, ComponentRecord>(
+            "SELECT components.namespace, components.name, component_versions.* FROM components JOIN component_versions ON components.component_id = component_versions.component_id WHERE components.namespace = $1 AND component_versions.component_id = $2"
```

Comment:
Could you please format long sql queries?

Reply:
formated